### PR TITLE
Add CORS support

### PR DESCRIPTION
**Pull request by boaz0**

### CODE REVIEW

File: cmd/podman/system/service_abi.go [hunk 1]
```diff
@@ -72,7 +72,7 @@ func restService(opts entities.ServiceOptions, flags *pflag.FlagSet, cfg *entiti
 	}
 
 	infra.StartWatcher(rt)
-	server, err := api.NewServerWithSettings(rt, opts.Timeout, listener)
+	server, err := api.NewServerWithSettings(rt, opts.Timeout, listener, opts.CorsHeaders)
```
Comment (rhatdan) [inline cmd/podman/system/service_abi.go:75]: Should we change the API to just pass in opts?

File: pkg/api/server/server.go [hunk 2]
```diff
@@ -34,10 +34,12 @@ type APIServer struct {
 	context.CancelFunc               // Stop APIServer
 	idleTracker        *idle.Tracker // Track connections to support idle shutdown
 	pprof              *http.Server  // Sidecar http server for providing performance data
+	CorsHeaders        string        // Inject CORS headers to each request
 }
 
 // Number of seconds to wait for next request, if exceeded shutdown server
 const (
+	DefaultCorsHeaders       = "*"
```
Comment (rhatdan) [inline pkg/api/server/server.go:42]: Should this be "*" or just ""?

**Review by rhatdan — APPROVED** [review]
A couple of questions.

File: pkg/api/server/server.go (hunk 2 above)
Comment (boaz0) [reply pkg/api/server/server.go:42]: I will change it, it should be empty string `""`

File: cmd/podman/system/service_abi.go (hunk 1 above)
Comment (boaz0) [reply cmd/podman/system/service_abi.go:75]: I thought about it and I think so. It will be easier.

File: cmd/podman/system/service.go [hunk 3]
```diff
@@ -39,7 +39,8 @@ Enable a listening service for API access to Podman commands.
 	}
 
 	srvArgs = struct {
-		Timeout int64
+		Timeout     int64
+		corsHeaders string
```
Comment (mheon) [inline cmd/podman/system/service.go:43]: I think this needs to be Capitalized (first letter, at least) so the value can be set by the CLI package

Comment (boaz0) [reply cmd/podman/system/service.go:43]: Right, good catch

**Comment by boaz0** [comment]
@rhatdan I am now figuring out a way to test this. It seems like I will have to come up with a solution from scratch.

File: pkg/api/server/handler_api.go [hunk 4]
```diff
@@ -63,6 +63,12 @@ func (s *APIServer) APIHandler(h http.HandlerFunc) http.HandlerFunc {
 			w.Header().Set("Libpod-API-Version", lv)
 			w.Header().Set("Server", "Libpod/"+lv+" ("+runtime.GOOS+")")
 
+			if s.CorsHeaders != "" {
+				w.Header().Set("Access-Control-Allow-Origin", s.CorsHeaders)
+				w.Header().Set("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, X-Registry-Auth")
```
Comment (jwhonce) [inline pkg/api/server/handler_api.go:68]: Do we need "Connection" and "Upgrade" in this header list?
We do need "X-Registry-Config".

**Review by jwhonce — CHANGES_REQUESTED** [review]
One question and one change request.

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM once tests and @jwhonce 's concerns are addressed

**Comment by TomSweeneyRedHat** [comment]
gack, hit the wrong button, didn't mean to close.

**Review by jwhonce — APPROVED** [review]

**Pull request merged**